### ISSUE TITLE
[Mochi] Permute op going crazy

### ISSUE THREAD

**Issue by vkovinicTT** [tenstorrent/tt-mlir#7729]
### Context

In the Mochi video generation model, the VAE decoder uses a [3D pixel shuffle](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/autoencoders/autoencoder_kl_mochi.py#L395) (depth-to-space-time) to upsample video frames. This is a standard technique where channel depth is traded for spatial/temporal resolution - no transposed convolutions, just a reshape-permute-reshape sequence:
```python
batch_size, num_channels, num_frames, height, width = hidden_states.shape
st = self.temporal_expansion    # 3, 2, or 1
sh = self.spatial_expansion     # always 2
sw = self.spatial_expansion     # always 2

hidden_states = hidden_states.view(batch_size, -1, st, sh, sw, num_frames, height, width)
hidden_states = hidden_states.permute(0, 1, 5, 2, 6, 3, 7, 4).contiguous()
hidden_states = hidden_states.view(batch_size, -1, num_frames * st, height * sh, width * sw)
```

This is what happens (using up_block_0 as an example, where `C_out=512, st=3, sh=2, sw=2, T=8, H=60, W=106`):

```python
# Input after linear projection that expanded channels: 768 -> 6144 (= 512 * 3 * 2 * 2)
# Shape: [1, 6144, 8, 60, 106]                          -- 596 MB (bf16)

# Step 1: Factor channel dim into (C_out, st, sh, sw, T, H, W)
x = x.view(1, 512, 3, 2, 2, 8, 60, 106)                # [B, C, st, sh, sw, T, H, W]
# Padded: last 2 dims (60, 106) -> (64, 128) = 1.29x    -- 768 MB, fine

# Step 2: Interleave factors with spatial dims
x = x.permute(0, 1, 5, 2, 6, 3, 7, 4)                  # [B, C, T, st, H, sh, W, sw]
# Shape: [1, 512, 8, 3, 60, 2, 106, 2]
# Padded: last 2 dims (106, 2) -> (128, 32) = 19.32x    -- 11,520 MB! <-- PROBLEM

# Step 3: Merge interleaved pairs into expanded spatial dims
x = x.view(1, 512, 24, 120, 212)                        # [B, C, T*st, H*sh, W*sw]
# Padded: last 2 dims (120, 212) -> (128, 224) = 1.13x  -- 672 MB, fine
```
### The problem

The permute at Step 2 produces an 8D tensor `[1, 512, 8, 3, 60, 2, 106, 2]` where the **last dimension is 2**. Since `getTilePaddedShape()` pads the last two dims to multiples of 32:

- `dim[-1] = 2` pads to `32` → **16x waste**
- `dim[-2] = 106` pads to `128` → 1.21x waste
- Combined: **19.32x overhead** — a 596 MB tensor becomes **11,520 MB (11.25 GB)**

This happens for all 3 pixel shuffles in the decoder. The second and third are even worse because the tensor is larger:

| Location | 8D permute output shape | Logical size | Tiled size | Overhead | Fits in 32 GB? |
|----------|------------------------|-------------|-----------|----------|----------------|
| up_block_0 | `[1,512,8,3,60,2,106,2]` | 596 MB | **11,520 MB** | 19.3x | Yes, but fragments memory |
| up_block_1 | `[1,256,24,2,120,2,212,2]` | 2,330 MB | **40,320 MB** | 16.9x | **No** |
| up_block_2 | `[1,128,48,1,240,2,424,2]` | 4,660 MB | **80,640 MB** | 16.9x | **No** |

The `up_block_1` permute alone requires 40 GB in TILE layout — it cannot physically fit on a single 32 GB Blackhole device, regardless of what else is in memory.

### Proposed solution: execute tile-hostile permutes in ROW_MAJOR layout

Since the root cause is tile padding on a tensor with a tiny trailing dimension, the fix is to keep the permute (and the reshape that follows it) in ROW_MAJOR layout. The tiling conversion (ROW_MAJOR → TILE) then happens on the well-shaped 5D output where the overhead is only 1.13x.

The layout cascade would be:

```
reshape [1,6144,8,60,106] -> [1,512,3,2,2,8,60,106]     TILE      768 MB   (fine)
permute -> [1,512,8,3,60,2,106,2]                        ROW_MAJOR 596 MB   (no tile padding!)
reshape -> [1,512,24,120,212]                             ROW_MAJOR 596 MB   (reshape preserves layout)
to_layout(ROW_MAJOR -> TILE)                              TILE      672 MB   (1.13x, fine)
```

We can make some condition on when to apply this, e.g. when tensor overhead is > 8GB TILE layout.

### Model-level workaround

We can make model-level workaround that avoids the problem entirely by decomposing the pixel shuffle into staged merges - one dimension pair at a time, keeping large dims in the tile-padded positions:

```python
# Original: one permute + one reshape (last dim = sw = 2, causes 16-19x overhead)
x = x.view(B, C, st, sh, sw, T, H, W)
x = x.permute(0, 1, 5, 2, 6, 3, 7, 4)             # [B, C, T, st, H, sh, W, sw]  <- sw=2 last!
x = x.view(B, C, T*st, H*sh, W*sw)

# Staged: 3 permutes + 2 reshapes (last 2 dims always large, max 1.33x overhead)
x = x.view(B, C, st, sh, sw, T, H, W)              # [B, C, st, sh, sw, T, H, W]
x = x.permute(0, 1, 5, 2, 3, 4, 6, 7)              # [B, C, T, st, sh, sw, H, W]  last2=(H,W)
x = x.reshape(B, C, T*st, sh, sw, H, W)             # merge (T,st)->T*st            last2=(H,W)
x = x.permute(0, 5, 3, 6, 4, 1, 2)                  # [B, H, sh, W, sw, C, T*st]   last2=(C,T*st)
x = x.reshape(B, H*sh, W*sw, C, T*st)               # merge (H,sh) and (W,sw)       last2=(C,T*st)
x = x.permute(0, 3, 4, 1, 2)                        # [B, C, T*st, H*sh, W*sw]      last2=(H*sh,W*sw)
```

**Comment by odjuricicTT** [comment]
@vkovinicTT Thanks! Can you expand on the example a bit more, what are some of the actual compute ops that come before and after this reshape / permute / reshape sequence? And what are their shapes?

**Comment by vkovinicTT** [comment]
@odjuricicTT here is more context regarding ops that come before and after pixel shuffle:

The pixel shuffle sits between a linear projection (`nn.Linear`) and a `GroupNorm` + `SiLU` + `Conv3d`. Here's the full PyTorch context from [`MochiUpBlock3D.forward()`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/autoencoders/autoencoder_kl_mochi.py#L367):

```python
# --- BEFORE: linear projection (768 -> 6144 channels) ---
hidden_states = hidden_states.permute(0, 2, 3, 4, 1)   # [1, 768, 8, 60, 106] -> [1, 8, 60, 106, 768]
hidden_states = self.proj(hidden_states)                 # nn.Linear(768, 6144) -> [1, 8, 60, 106, 6144]
hidden_states = hidden_states.permute(0, 4, 1, 2, 3)   # -> [1, 6144, 8, 60, 106]

# --- PIXEL SHUFFLE (the problematic part) ---
hidden_states = hidden_states.view(B, -1, st, sh, sw, T, H, W)          # [1, 512, 3, 2, 2, 8, 60, 106]
hidden_states = hidden_states.permute(0, 1, 5, 2, 6, 3, 7, 4)          # [1, 512, 8, 3, 60, 2, 106, 2]  <-- problem
hidden_states = hidden_states.view(B, -1, T*st, H*sh, W*sw)             # [1, 512, 24, 120, 212]

# --- AFTER: next ResBlock starts with GroupNorm + SiLU + Conv3d(512, 512, 3x3x3) ---
# hidden_states = self.norm1(hidden_states)     # MochiChunkedGroupNorm3D(512, groups=32)
# hidden_states = F.silu(hidden_states)
# hidden_states = self.conv1(hidden_states)     # CogVideoXCausalConv3d(512, 512, kernel_size=3)
```

And here's the corresponding compiled IR (from the log):
```mlir
// BEFORE: linear projection (SyncTensorsGraph.20)
%2 = "ttir.permute"(%1) <{permutation = array<i64: 1, 0>}>
     : (tensor<6144x768xbf16>) -> tensor<768x6144xbf16>                         // transpose weight
%3 = "ttir.dot_general"(%arg2, %2) <{contract_dims_lhs = [4], contract_dims_rhs = [0]}>
     : (tensor<1x8x60x106x768xbf16>, tensor<768x6144xbf16>)                    // loc("aten__einsum")
     -> tensor<1x8x60x106x6144xbf16>
%8 = "ttir.add"(%3, bias_broadcast) -> tensor<1x8x60x106x6144xbf16>            // loc("aten__add")
// then permuted NDHWC -> NCDHW in SyncTensorsGraph.5:
// "ttir.permute" dims=[0,4,1,2,3] : -> tensor<1x6144x8x60x106xbf16>

// PIXEL SHUFFLE
%0 = "ttir.reshape"(%input) <{shape = [1, 512, 3, 2, 2, 8, 60, 106]}>
     : (tensor<1x6144x8x60x106xbf16>) -> tensor<1x512x3x2x2x8x60x106xbf16>
     // last 2 dims = (60, 106) -> tile-padded (64, 128) = 768 MB, 1.29x overhead — OK

%1 = "ttir.permute"(%0) <{permutation = array<i64: 0, 1, 5, 2, 6, 3, 7, 4>}>
     : (tensor<1x512x3x2x2x8x60x106xbf16>) -> tensor<1x512x8x3x60x2x106x2xbf16>
     // last 2 dims = (106, 2) -> tile-padded (128, 32) = 11,520 MB, 19.3x overhead — PROBLEM
     // memref layout: <5898240x1 x tile<32x32, bf16>>  (5.9M single-column tiles)

%2 = "ttir.reshape"(%1) <{shape = [1, 512, 24, 120, 212]}>
     : (tensor<1x512x8x3x60x2x106x2xbf16>) -> tensor<1x512x24x120x212xbf16>
     // last 2 dims = (120, 212) -> tile-padded (128, 224) = 672 MB, 1.13x overhead — OK

// AFTER: GroupNorm (chunked over temporal dim, 3 slices of 8 frames each)
// ttnn.permute  1x512x24x120x212 -> 1x24x512x120x212
// ttnn.reshape  1x24x512x120x212 -> 24x512x120x212
// ttnn.slice    24x512x120x212   -> 8x512x120x212      (temporal chunk)
// ttnn.reshape  8x512x120x212    -> 8x32x16x25440      (32 groups)
// ttnn.sum      8x32x16x25440    -> 8x32               (mean for var_mean)
// ... (normalize, affine, concat chunks back)
//
// Then SiLU + Conv3d:
// ttnn.silu     1x512x24x120x212 -> 1x512x24x120x212
// ttnn.conv3d   1x26x122x214x512 @ 13824x128 -> 1x24x120x212x128   (512->512, 3x3x3)
```
The problem is isolated to the single `ttir.permute` that puts `sw=2` in the last dimension.
